src/Core/Resolvers/MultipleCreateOrderHelper.cs

```diff
@@ -0,0 +1,359 @@
+                // and there is an autogenerated relationship field in the source entity which suggests us to perform the insertion first in source entity,
+                // we cannot determine a valid order of insertion.
+                if (doesTargetBodyContainAnyRelationshipField && doesSourceContainAnyAutogenRelationshipField)
+                {
+                    throw new DataApiBuilderException(
+                        message: $"Input for target entity: {targetEntityName} for the relationship: {relationshipName} at level: {nestingLevel} cannot contain the field: {relationshipFieldInSource}.",
+                        statusCode: HttpStatusCode.BadRequest,
+                        subStatusCode: DataApiBuilderException.SubStatusCodes.BadRequest);
+                }
+
+                // Determine whether the input data for source/target contain a value (could be null) for this pair of relationship fields. 
+                bool doesSourceBodyContainThisRelationshipField = columnDataInSourceBody.TryGetValue(relationshipFieldInSource, out IValueNode? sourceColumnvalue);
+                bool doesTargetBodyContainThisRelationshipField = columnDataInTargetBody.TryGetValue(relationshipFieldInTarget, out IValueNode? targetColumnvalue);
+
+                // Update whether input data for source/target contains any relationship field.
+                doesSourceBodyContainAnyRelationshipField = doesSourceBodyContainAnyRelationshipField || doesSourceBodyContainThisRelationshipField;
+                doesTargetBodyContainAnyRelationshipField = doesTargetBodyContainAnyRelationshipField || doesTargetBodyContainThisRelationshipField;
+
+                // If relationship columns are present in the input for both the source and target entities,
+                // we cannot choose one entity as the referencing entity. This is because for a referencing entity,
+                // the values for all the referencing fields should be derived from the insertion in the referenced entity.
+                // However, here both entities contain atleast one relationship field whose value is provided in the request.
+                if (doesSourceBodyContainAnyRelationshipField && doesTargetBodyContainAnyRelationshipField)
+                {
+                    throw new DataApiBuilderException(
+                        message: $"The relationship fields must be specified for either the source entity: {sourceEntityName} or the target entity: {targetEntityName} " +
+                        $"for the relationship: {relationshipName} at level: {nestingLevel}.",
+                        statusCode: HttpStatusCode.BadRequest,
+                        subStatusCode: DataApiBuilderException.SubStatusCodes.BadRequest);
+                }
+
+                // The source/target entities can assume a value for insertion for a relationship field if:
+                // 1. The field is autogenerated, or
```

Comment:
How about when a relationship field is backed by a column with a default value?

*Scenario where it should be possible to determine the referenced/referencing entity:*

Consider two entities: 
Entity A ---> Relationship fields: [A1, A2]. Field A2 is backed by a column with a default value.  
Entity B ---> Relationship fields: [B1, B2]. 

In the mutation request, if there is data for A1 and no data for A2, B1, B2; it should be possible for us to say Entity B is the referencing entity since, insertion in Entity A will succeed because we have values for A1, A2 (and rest of the columns).  As long as the combination of <A1, A2> is unique, it should be possible to use this to link it with a row in <B1,B2>


*Scenarios where it is not possible to determine the referencing/referenced entity:*

Consider two entities: 
Entity A ---> Relationship fields: [A1, A2]. Field A2 is backed by a column with a default value.  
Entity B ---> Relationship fields: [B1, B2].  Field B2 is backed by a column with a default value.

In the mutation request, if there is data for A1, B1 and no data for A2 and B2; Since, both A2 and B2 are backed by columns with default values, insertion in either one can succeed. It is not possible to determine which is the referencing entity here. 

IMHO, evaluating different scenarios related to a field backed by a column with default value can be taken up as a follow-up issue/PR.

Reply:
Created this issue to track: https://github.com/Azure/data-api-builder/issues/2148